### DISCUSSION TITLE
Composition Gradient - Material Property Gradient

### DISCUSSION THREAD

**Discussion by nsundquist** [idaholab/moose#20171] [Q&A General] [answered]
Hi all,

I'm currently working on a phase separation problem and I need to use the x component of the gradient of the composition. I've been trying to use a DerivativeParsedMaterial to generate the x derivative, but this just outputs zero. 

```
[dc2dx_mat]
    type = DerivativeParsedMaterial
    f_name = dc2dx_mat
    material_property_names = 'dc2dx:=D[c2,x]'
    function = 'dc2dx'
    outputs = exodus
    derivative_order = 2
  []
```

The SplitCHParsed kernel is used to generate the values of c2.

```
 [c2_res]
    type = SplitCHParsed
    variable = c2
    f_name = F_sum
    kappa_name = kappa_c
    w = w2
    args = 'c3'
  []
```

Thank you for any help you can offer.

**Comment by GiudGiud** [comment]
Hello

I'm not sure we can use the DerivativeParsedMaterial for a gradient, unless maybe we define 'x' to be a material property. @dschwen will know
I'd just use this to get the gradient component. 
https://mooseframework.inl.gov/source/auxkernels/VariableGradientComponent.html

Guillaume

**Reply by nsundquist** [reply]
Thank you for your response and sorry it took me so long to reply.

I've tried using VariableGradientComponent, but I get this error: "nodal variables do not have gradients". I'm not sure what makes c2 a nodal variable.

**Reply by GiudGiud** [reply]
A nodal variable is defined by its value at nodes.
the issue here is that the shape function derivatives are discontinuous at nodes, so the gradient is not well defined 

You should project the variable to an elemental one, then get the gradient of that variable. It wont be exact though. Is this just for postprocessing?

**Reply by nsundquist** [reply]
I need to use the gradient to calculate the demagnetizing field.

**Comment by permcody** [comment]
Which shape function family and order are you using? Note that linear Lagrange doesn't support a second derivative.

**Reply by nsundquist** [reply]
Thank you so much for responding, I'm sorry it took me this long to reply. 

I've tried using first and second order Lagrange, both gave the same output of zero. I also tried using the Hermite family and got a segmentation fault. What function family would you recommend using for this application?

**Comment by GiudGiud** [comment, marked as answer]
See https://github.com/idaholab/moose/discussions/20628 and https://github.com/idaholab/moose/discussions/18674 for more discussion on this. 
Top solutions for short:
- use functor material properties instead of regular material properties
OR
- write the gradient by deriving it yourself, defining a new material property
